Content.Shared/Access/Components/AccessComponent.cs

```diff
@@ -57,3 +57,17 @@ public void AddGroup(ProtoId<AccessGroupPrototype> group)
         Tags.UnionWith(groupPrototype.Tags);
     }
 }
+
+/// <summary>
+/// Event raised on an entity attempting to interact with something that is access restricted.
+/// </summary>
+[ByRefEvent]
+public sealed class AccessAttemptedEvent : CancellableEntityEventArgs
+{
+    public EntityUid? User;
+
+    public AccessAttemptedEvent(EntityUid? user)
+    {
+        User = user;
+    }
+}
```

Comment:
GunSystem should just be calling IsAllowed but inside the TryX methods, this is just an event wrapping a method call atm.